identifier.go

```diff
@@ -80,24 +79,24 @@ func encodeIdentifier(identifier Identifier) []byte {
 
 		tag := identifier.Tag
 
-		highBit := uint(63)
-		for {
-			if tag&(1<<highBit) != 0 {
-				break
-			}
-			highBit--
-		}
+		b = append(b, encodeHighTag(tag)...)
+	}
+	return b
+}
 
-		tagBytes := int(math.Ceil(float64(highBit) / 7.0))
-		for i := tagBytes - 1; i >= 0; i-- {
-			offset := uint(i) * 7
-			mask := Tag(0x7f) << offset
-			tagByte := (tag & mask) >> offset
-			if i != 0 {
-				tagByte |= 0x80
-			}
-			b = append(b, byte(tagByte))
+func encodeHighTag(tag Tag) []byte {
+	b := make([]byte, 0, 4)
+	for tag != 0 {
```

Comment:
This seems like a cleaner implementation (and similar to other implementations I've seen in python for example)  I'll have to take some time to actually understand it though.  I trust the tests are robust enough to demonstrate that this is at least as good as the old implementation.

Reply:
@johnweldon This part is the bug fix. I didn't dive deep on the previous implementation but it looked like an OBOE. The [Travis output](https://travis-ci.org/go-asn1-ber/asn1-ber/jobs/279164682#L506-L512) shows it munching away bytes in the unpack-pack-unpack operations.

```
--- FAIL: TestSuiteDecodePacket (0.00s)
	suite_test.go:118: tests/tc49.ber: data should be the same
		want: []byte{0x9f, 0x81, 0x6, 0x0}
		got: []byte{0x9f, 0x6, 0x0}
	suite_test.go:130: tests/tc49.ber: data should be the same
		want: []byte{0x9f, 0x6, 0x0}
		got: []byte{0x86, 0x0}
```

Here's an annotated version of `encodeHighTag`:

```go
func encodeHighTag(tag Tag) []byte {
	// set cap=4 so hopefully we don't have to do another allocation
	b := make([]byte, 0, 4)
	for tag != 0 {
		// t := last 7 bits of tag (HighTagValueBitmask = 0x7F)
		t := tag & HighTagValueBitmask

		// rshift tag 7 to remove what we just pulled off
		tag >>= 7

		// if we previously added to b this entry needs a continuation bit (0x80)
		if len(b) != 0 {
			// add the continuation bit
			t |= HighTagContinueBitmask
		}

		b = append(b, byte(t))
	}
	// reverse
	// since we pulled bits off tag small to high our byte slice is in the wrong order
	// example: tag = 0xFF results in {0x7F, 0x01 + 0x80 (continuation bit)}
	// this needs to be reversed into 0x81 0x7F
	for i, j := 0, len(b)-1; i < len(b)/2; i++ {
		b[i], b[j-i] = b[j-i], b[i]
	}
	return b
}
```

Reply:
+1 for the annotated version... can you add the comments to the commit?